lib/hipcore_typed/syntax.ml

```diff
@@ -17,16 +17,25 @@ let disj = Lists.foldr1 (fun c t -> Disjunction (c, t))
 let conj = Lists.foldr1 ~default:True (fun c t -> And (c, t))
 let sep_conj = Lists.foldr1 ~default:EmptyHeap (fun c t -> SepConj (c, t))
 
-let eq x y = Atomic (EQ, x, y)
+let term term_desc term_type = {term_desc; term_type}
+
+let eq x y =
+  let x, y = 
+    let value_type = match x.term_type, y.term_type with
+    | Any, t | t, Any -> t
+    (* anything else needs an inference context to unify, make a heuristic guess *)
+    | _, _ -> x.term_type
```

Comment:
Hmm, this seems like it will bite us in future... are there cases where we need to add equalities between things with types which aren't completely known? Should there be a case for when the types are equal?